### PR TITLE
feat: add sort-regexp rule

### PR DESCRIPTION
**Pull request by azat-io**
### Description

A new rule for sorting regular expressions is added. It works with:

- capture groups (`/(error|warning|info|debug)/`)
- character classes (`/[zxa-f0-9A-Z]/`)
- flags (`/pattern/igmus`)

### Additional context

N/A.

---

### What is the purpose of this pull request? 

- [ ] Bug fix
- [x] New Feature
- [ ] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves.
- [x] Ideally, include relevant tests that fail without this PR but pass with it.
- [x] Read [contribution guidelines](https://github.com/azat-io/eslint-plugin-perfectionist/blob/main/contributing.md).

### CODE REVIEW

File: test/rules/sort-regexp.test.ts [hunk 1]
```diff
@@ -0,0 +1,2835 @@
+/* Cspell:ignore gimsu igmus yusmig ysumgi Ωmega Δelta */
```
[the remark this answers is not in the document]
Comment (azat-io) [reply test/rules/sort-regexp.test.ts]: 🤦

File: docs/content/rules/sort-regexp.mdx [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+import CodeExample from '../../components/CodeExample.svelte'
+import CodeTabs from '../../components/CodeTabs.svelte'
+import dedent from 'dedent'
+
+Enforce consistent ordering in regular expressions: flags, character-class elements, and alternation branches inside capture groups or the top-level pattern.
+
+Large patterns become easier to maintain when related branches are grouped and repeatedly used flags or character classes follow the same order. This rule helps you spot missing alternatives, avoid duplicated work, and keeps refactors safer by applying one sorting strategy across your codebase.
+
+The rule is **safe** – it only reorders elements without changing their meaning.
+
```
[the remark this answers is not in the document]
Comment (hugop95) [reply docs/content/rules/sort-regexp.mdx:23]: @azat-io 

There are indeed dangers with the current state:

image

but

image

---

I don't think we should include this rule in the `recommended`, they are very likely other edge cases we will be missing.

[the remark this answers is not in the document]
Comment (azat-io) [reply docs/content/rules/sort-regexp.mdx:23]: @hugop95 I fixed it

File: test/rules/sort-regexp.test.ts [hunk 3]
```diff
@@ -0,0 +1,2835 @@
...
+          /[a\x61]/
+        `),
+        options: [customOptions],
+      })
+    })
+
+    it('sorts character classes', async () => {
+      let customOptions = {
+        ...options,
+        type: 'natural',
```
[the remark this answers is not in the document]
Comment (azat-io) [reply test/rules/sort-regexp.test.ts]: Fixed.

File: rules/sort-regexp.ts [hunk 4]
```diff
@@ -617,6 +647,22 @@ function getAlternativeAlias(alternative: Alternative): string | null {
...
+  if (first.length === 0 || second.length === 0) {
+    return true
+  }
+
+  if (first.length === second.length) {
+    return first === second
+  }
+
+  if (first.length < second.length) {
+    return second.startsWith(first)
```
Comment (hugop95) [inline rules/sort-regexp.ts]: @azat-io 

This won't work for cases such as: `/(a|.*ab)/.exec("ab")` and `/(.*ab|a)/.exec("ab")`.

```ts
/(a|.*ab)/.exec("ab")
(2) ['a', 'a', index: 0, input: 'ab', groups: undefined]

/(.*ab|a)/.exec("ab")
(2) ['ab', 'ab', index: 0, input: 'ab', groups: undefined]
```

I haven't looked deep into it, but we probably need to `includes` for maximal safety.

In fact, it simply might never be fully safe with `exec`.

Maybe we would have to restrain sorting to specific usages (`.test` for example).

File: docs/content/rules/sort-regexp.mdx [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+import CodeExample from '../../components/CodeExample.svelte'
+import CodeTabs from '../../components/CodeTabs.svelte'
+import dedent from 'dedent'
+
+Enforce consistent ordering in regular expressions: flags, character-class elements, and alternation branches inside capture groups or the top-level pattern.
+
+Large patterns become easier to maintain when related branches are grouped and repeatedly used flags or character classes follow the same order. This rule helps you spot missing alternatives, avoid duplicated work, and keeps refactors safer by applying one sorting strategy across your codebase.
+
+The rule is **safe** – it only reorders elements without changing their meaning.
```
Comment (hugop95) [inline docs/content/rules/sort-regexp.mdx:22]: @azat-io 

Honestly, I'm not confident that we can 100% ensure full safety here at this stage. I would even say that regex re-ordering probably has more weird edge cases that any other rules we have (which already have complex edge cases due to dependency ordering).

I expect bugs related to ordering that change the behavior when the rule is launched. It might be OK in 95% of the cases, but I have doubt about the remaining complex cases.

File: rules/sort-regexp/get-sorting-node-name.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ * @param parameters - Alternative metadata and rule options.
+ * @returns String representation used during sorting.
+ */
+export function getSortingNodeName({
+  alternativeAlias,
+  alternative,
+  options,
+}: SortingNodeNameParameters): string {
+  if (!options.ignoreAlias && alternativeAlias) {
+    return `${alternativeAlias}: ${alternative.raw}`
```
Comment (hugop95) [inline rules/sort-regexp/get-sorting-node-name.ts]: @azat-io 

I would remove the space after `:`.

Comment (azat-io) [reply rules/sort-regexp/get-sorting-node-name.ts]: @hugop95 Quick question about the space after `:` - you suggested removing it from `"pet: cat"` → `"pet:cat"`.

Is this a style preference, or is there a specific reason? I kept the space for readability in error messages, but happy to change if there's a convention I'm missing.

File: docs/content/rules/sort-regexp.mdx (hunk 5 above)
Comment (azat-io) [reply docs/content/rules/sort-regexp.mdx:22]: I've added additional safeguards in the latest commits:

  - **ES2024 `/v` flag guard** - skips character class sorting to avoid breaking set operations (`&&`, `--`)
  - **More backreference tests** - covers named groups and complex scenarios
  - **Existing shadowing detection** - already handles `/(ab|a)/` cases

The rule is conservative - it skips sorting when it detects potentially unsafe patterns. That said, I agree 100% safety is hard to guarantee with regex.

**Comment by azat-io** [comment]
I’m currently testing the rule on large codebases.

So far I haven't managed to fix the issue where line breaks are split or checked.

- Before: `/\r\n|[\r\n\u2028\u2029]/`
- After: `/[\n\r\u2028\u2029]|\r\n/`

In JavaScript, alternatives are tried from left to right. When \r\n is moved behind the character class, the engine will match the single `\r` first, leaving the `\n` unmatched.

This matters in real code:

```js
const LINEBREAK_MATCHER_OLD = /\r\n|[\r\n\u2028\u2029]/;
const LINEBREAK_MATCHER_NEW = /[\n\r\u2028\u2029]|\r\n/;

let text = 'a\r\nb';

console.log(text.split(LINEBREAK_MATCHER_OLD)); // ['a', 'b']
console.log(text.split(LINEBREAK_MATCHER_NEW)); // ['a', '', 'b']
```

**Comment by azat-io** [comment]
I fixed the issue mentioned above and added the necessary tests.

I also tested the new ESLint rule on several large projects:

- [Babel](https://github.com/babel/babel)
- [ESLint](https://github.com/eslint/eslint)
- [TypeScript ESLint](https://github.com/typescript-eslint/typescript-eslint)
- [Vue](https://github.com/vuejs/core)
- [Astro](https://github.com/withastro/astro)
- [Prettier](https://github.com/prettier/prettier)

All changes and auto-fixes are cosmetic in nature and do not create functional changes.

@hugop95 This PR is now ready for review.

File: rules/sort-regexp.ts [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+
+          if (
+            hasShadowingAlternatives({
+              alternatives: alternative.parent.alternatives,
+              flags: literalNode.regex.flags,
+            })
+          ) {
+            literalsWithShadowing.add(literalNode)
+            return
+          }
```
[the remark this answers is not in the document]
Comment (azat-io) [reply rules/sort-regexp.ts:201]: `hasShadowingAlternatives` already treats `/\r\n|[\r\n\u2028\u2029]/` as unsafe, so the rule marks that literal as shadowed and leaves both the disjunction and the character class untouched. This behavior is covered by `test/rules/sort-regexp-shadowing.test.ts:405` and `test/rules/sort-regexp.test.ts:588`

File: docs/content/rules/sort-regexp.mdx [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+<sub>default: `false`</sub>
+
+Determines whether named capturing group aliases (e.g., `(?<role>...)`) are considered during comparison.
+
+- `false` — Sort by `alias: pattern` (default).
+- `true` — Ignore the alias and only compare the branch content.
+
+This is useful when you only care about the pattern and not how it is aliased.
+
+### customGroups
```
Comment (hugop95) [inline docs/content/rules/sort-regexp.mdx:137]: @azat-io 

For other rules, this section is after `Groups`, so I would do the same here for consistency.

File: docs/content/rules/sort-regexp.mdx [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+- `selector` — Filter by element type:
+  - `'alias'` — Named capturing group alternatives.
+  - `'pattern'` — Plain alternatives without aliases.
+- `elementNamePattern` — Regex applied to the alias (if present).
+- `elementValuePattern` — Regex applied to the branch content.
+- `type`, `order`, `fallbackSort` — Override the respective global options for the group.
+
+Custom groups are evaluated in order; the first match wins and overrides predefined groups.
+
+### groups
```
Comment (hugop95) [inline docs/content/rules/sort-regexp.mdx:182]: @azat-io 

I would take the same `sub` sections as for other rules for consistency as well, even if this feature probably won't likely be used for this rule in particular:

https://github.com/azat-io/eslint-plugin-perfectionist/blob/f9919cb36d64c5acb7d41723e874bb316e84b452/docs/content/rules/sort-objects.mdx?plain=1#L460-L477

File: package.json [hunk 10]
```diff
@@ -57,6 +57,7 @@
     "test:unit": "vitest --run --coverage"
   },
   "dependencies": {
+    "@eslint-community/regexpp": "^4.12.1",
```
Comment (hugop95) [inline package.json:60]: @azat-io 

For information:
- `@eslint-community/regexpp`'s unpacked size is **474kB** according to [NPM](https://www.npmjs.com/package/@eslint-community/regexpp):
- [Perfectionist V4](https://www.npmjs.com/package/eslint-plugin-perfectionist) is at **327kb** 
- [`@typescript-eslint/utils`](https://www.npmjs.com/package/@typescript-eslint/utils) is at **195kB**.
- [`natural-orderby`](https://www.npmjs.com/package/natural-orderby) is at **72kb**.

Just something to keep in mind based on the adoption of the rule.

File: rules/sort-regexp.ts [hunk 11]
```diff
@@ -0,0 +1,416 @@
...
+import { reportAllErrors } from '../utils/report-all-errors'
+import { ORDER_ERROR } from '../utils/report-errors'
+import { getSettings } from '../utils/get-settings'
+import { isSortable } from '../utils/is-sortable'
+import { sortNodes } from '../utils/sort-nodes'
+import { pairwise } from '../utils/pairwise'
+import { complete } from '../utils/complete'
+import { compare } from '../utils/compare'
+
+type SortRegExpSortingNode = SortingNode<TSESTree.Literal>
```
Comment (hugop95) [inline rules/sort-regexp.ts:42]: @azat-io 

We could put and export this type in `rules/sort-regexp/types.ts`, and use it in places doing `SortingNode<TSESTree.Literal>` again.
- In `createCharacterClassSortingNode` in this file.
- `rules/sort-regexp/create-flag-sorting-nodes.ts`.
- `rules/sort-regexp/create-character-class-sorting-node.ts`.

File: rules/sort-regexp.ts [hunk 12]
```diff
@@ -0,0 +1,416 @@
...
+
+type MessageId = 'unexpectedRegExpOrder'
+
+/**
+ * Retrieves the original source text for a character class element.
+ *
+ * @param parameters - Character class element metadata.
+ * @returns Slice of the literal raw string covering the element.
+ */
+function getCharacterClassElementRawText({
```
Comment (hugop95) [inline rules/sort-regexp.ts:54]: @azat-io 

[nit] It's a bit strange IMO to have this local function at the top. What about putting it at the very bottom ?

File: rules/sort-regexp.ts [hunk 13]
```diff
@@ -0,0 +1,416 @@
...
+              sortingNode.node === sortedAlternatives.at(index)?.node,
+          )
+
+          if (isAlreadySorted) {
+            return
+          }
+
+          reportAllErrors<MessageId, SortRegExpSortingNode>({
+            availableMessageIds: {
+              unexpectedGroupOrder: 'unexpectedRegExpOrder',
```
Comment (hugop95) [inline rules/sort-regexp.ts:239]: @azat-io 

Could we create a dedicated `unexpectedRegExpGroupOrder` messageID to stay consistent with other rules ?

File: test/rules/sort-regexp.test.ts [hunk 14]
```diff
@@ -0,0 +1,3267 @@
...
+        output: dedent`
+          /(?<b>(?<a>a)|b)/
+        `,
+        code: dedent`
+          /(?<b>b|(?<a>a))/
+        `,
+      })
+    })
+
+    it('respects custom groups by alias name', async () => {
```
Comment (hugop95) [inline test/rules/sort-regexp.test.ts:1945]: @azat-io 

I think that the `groups` feature doesn't work: if you change `groups` to `["unknown", "digits"]`, the result is still valid.

The only case that uses `sortNodesByGroups` is for `sortAlternatives`.

**Review by hugop95 — COMMENTED** [review]
@azat-io 

This is a very ambitious initiative !

The PR is very large, so I haven't reviewed everything yet, but I have some comments already on a couple of things.

One problem that I think is blocking right now is that the `groups` feature doesn't work (or at the very least, tests do not test it correctly, I've left a comment for that).

My other feedback is regarding the current approach: the rule sorts 3 things as you mention in the description:
- capture groups `(/(error|warning|info|debug)/)`
- character classes `(/[zxa-f0-9A-Z]/)`
- flags `(/pattern/igmus)`

I wonder if it wouldn't have been better to go with 3 separate rules:
- They are all linked to RegExp, but represent very different things.
- We already have one rule that sorts multiple things at the same time: `sort-switch-case` (it can sort "groups" of `cases`, and what is inside a consecutive list of `cases`).
  - It's a bit of our ugly duckling rule: it's the only one that doesn't supports `groups`, because this notion mean different things depending on what we want to sort. 
  - We have a bit of the same case here:
    - What do `groups`/`customGroups` apply to ? `flags`, `capture groups`, or `character classes` ?
    - Can users `group` flags only, but not other things ?
    - I think that in all cases, this should be documented, because it's impossible to know otherwise.
- Code wise, it's harder to understand the sort of all 3 things in the same rule.
  - Maybe we could have common functions use by the 3 rules, but each rule is kept small and only focuses on sorting one thing. 
- `ignoreAlias` is only useful for `capture groups` sorting.
- This way, it allows users might to enable `flags` sorting, but not `capture groups`.
  - I think that it's important: flag sorting is **fully safe**, but capture groups isn't in theory.

Let me know what you think !

**Comment by azat-io** [comment]
Closed with this reason: https://github.com/azat-io/eslint-plugin-perfectionist/issues/566#issuecomment-3568080700
